src/ejabberd_receiver.erl

```diff
@@ -445,3 +445,19 @@ hibernate() ->
 -spec maybe_hibernate(state()) -> hibernate | infinity | pos_integer().
 maybe_hibernate(#state{hibernate_after = 0}) -> hibernate();
 maybe_hibernate(#state{hibernate_after = HA}) -> HA.
+
+%% It's workaround taken from ssl application
+workaround_transport_delivery_problems(Socket, gen_tcp) ->
```

Comment:
I think we can rename the function to sth like `ensure_data_sent_before_close`

Reply:
Renamed to shutdown_socket_and_wait_for_peer_to_close.
It does not ensure data is sent, it's *trying* to ensure that data is sent. It also closes socket (ok, peer closes).